File: _extensions/toggle/toggle.lua
--- Toggle Extension for Quarto
-- Provides toggle buttons to show/hide code output sections in Quarto HTML documents.
-- This Lua filter processes code cells and adds toggle functionality based on
-- document-level and cell-level configuration options.
--
-- @author James J Balamuta
-- @copyright 2025 James J Balamuta
-- @release 0.2.0
--
-- @usage
-- Document-level configuration:
--   extensions:
--     toggle:
--       output-toggle: true     # Enable toggle buttons (default: false)
--       output-hidden: true     # Hide outputs by default (default: false)
--       output-sync: true       # Sync all outputs in a cell (default: false)
--       global-toggle: true     # Show global toggle button (default: false)
--       persist: true           # Persist state in localStorage (default: false)
--       button-text: "Result"   # Custom button text (default: "Output")
--
-- Cell-level configuration:
--   ```{python}
--   #| toggle: true
--   #| output-hidden: true
--   print("Hello")
--   ```

-- ============================================================================
-- Module State
-- ============================================================================

--- Flag indicating if toggle dependencies are needed.
-- @field needs_toggle boolean
local needs_toggle = false

--- Cell counter for generating unique toggle IDs.
-- @field cell_index number Current cell index (0-based)
local cell_index = 0

--- Default settings values.
-- @class table
-- @name DEFAULT_SETTINGS
-- @field toggle_enabled boolean Default: false
-- @field output_hidden boolean Default: false
-- @field output_sync boolean Default: false
-- @field global_toggle boolean Default: false
-- @field persist boolean Default: false
-- @field button_text string Default: "Output"
local DEFAULT_SETTINGS = {
  toggle_enabled = false,
  output_hidden = false,
  output_sync = false,
  global_toggle = false,
  persist = false,
  button_text = "Output"
}

--- Document-level settings (populated from metadata).
-- @class table
-- @name doc_settings
local doc_settings = {}

--- Initialize doc_settings with default values.
-- Copies DEFAULT_SETTINGS to doc_settings.
-- @return nil
local function init_doc_settings()
  for k, v in pairs(DEFAULT_SETTINGS) do
    doc_settings[k] = v
  end
end

-- Initialize on load
init_doc_settings()

-- ============================================================================
-- Type Conversion Utilities
-- ============================================================================

--- Convert a value to boolean.
-- Safely converts Pandoc metadata values to Lua booleans.
-- @param value any The value to convert
-- @return boolean|nil The boolean representation or nil if input is nil
local function to_boolean(value)
  if value == nil then
    return nil
  end
  if type(value) == "boolean" then
    return value
  end
  return value == true
end

--- Convert a value to string.
-- Safely converts Pandoc metadata values to Lua strings.
-- @param value any The value to convert
-- @return string|nil The string representation or nil if empty/invalid
local function to_string(value)
  if value == nil then
    return nil
  end
  local str = pandoc.utils.stringify(value)
  if str and str ~= "" then
    return str
  end
  return nil
end

-- ============================================================================
-- Configuration Extraction
-- ============================================================================

--- Extract a single boolean setting from a config table.
-- @param config table The configuration table
-- @param key string The key to extract
-- @param current_value boolean The current/default value
-- @return boolean The extracted value or current value if not found
local function extract_bool_setting(config, key, current_value)
  if config[key] ~= nil then
    local val = to_boolean(config[key])
    if val ~= nil then
      return val
    end
  end
  return current_value
end

--- Extract a single string setting from a config table.
-- @param config table The configuration table
-- @param key string The key to extract
-- @param current_value string The current/default value
-- @return string The extracted value or current value if not found
local function extract_string_setting(config, key, current_value)
  if config[key] ~= nil then
    local val = to_string(config[key])
    if val then
      return val
    end
  end
  return current_value
end

--- Process a toggle configuration table and update settings.
-- Extracts all toggle options from a configuration table.
-- @param config table The configuration table (from YAML)
-- @param settings table The settings table to update
-- @return table The updated settings table
local function process_config_table(config, settings)
  settings.toggle_enabled = extract_bool_setting(config, "output-toggle", settings.toggle_enabled)
  settings.output_hidden = extract_bool_setting(config, "output-hidden", settings.output_hidden)
  settings.output_sync = extract_bool_setting(config, "output-sync", settings.output_sync)
  settings.global_toggle = extract_bool_setting(config, "global-toggle", settings.global_toggle)
  settings.persist = extract_bool_setting(config, "persist", settings.persist)
  settings.button_text = extract_string_setting(config, "button-text", settings.button_text)
  return settings
end

--- Create a new settings table with default values.
-- @return table A new settings table initialized with defaults
local function create_default_settings()
  local settings = {}
  for k, v in pairs(DEFAULT_SETTINGS) do
    settings[k] = v
  end
  return settings
end

--- Extract toggle settings from document metadata.
-- Parses both new format (extensions.toggle.*) and legacy format (toggle.*).
-- @param meta table The document metadata from Pandoc
-- @return table A settings table with all toggle configuration options
local function extract_settings(meta)
  local settings = create_default_settings()
  
  -- New format: extensions.toggle
  if meta.extensions and meta.extensions.toggle then
    settings = process_config_table(meta.extensions.toggle, settings)
  end
  
  -- Legacy format: toggle.* (processed second to allow override)
  if meta.toggle then
    settings = process_config_table(meta.toggle, settings)
  end
  
  return settings
end

-- ============================================================================
-- Cell Attribute Utilities
-- ============================================================================

--- Get a boolean attribute from a Div element.
-- @param el table The Pandoc Div element
-- @param attr_name string The attribute name
-- @param default boolean The default value
-- @return boolean The attribute value or default
local function get_bool_attr(el, attr_name, default)
  local value = el.attributes[attr_name]
  if value == nil then
    return default
  end
  return value == "true"
end

--- Get a string attribute from a Div element.
-- @param el table The Pandoc Div element
-- @param attr_name string The attribute name
-- @param default string The default value
-- @return string The attribute value or default
local function get_string_attr(el, attr_name, default)
  local value = el.attributes[attr_name]
  if value == nil or value == "" then
    return default
  end
  return value
end

--- Check if an attribute exists on a Div element.
-- @param el table The Pandoc Div element
-- @param attr_name string The attribute name
-- @return boolean True if attribute exists
local function has_attr(el, attr_name)
  return el.attributes[attr_name] ~= nil
end

-- ============================================================================
-- Cell Settings
-- ============================================================================

--- Create cell settings from document settings and element attributes.
-- Cell-level attributes override document-level settings.
-- @param el table The Pandoc Div element
-- @return table The resolved cell settings
local function resolve_cell_settings(el)
  local settings = {
    toggle = doc_settings.toggle_enabled,
    hidden = doc_settings.output_hidden,
    sync = doc_settings.output_sync,
    persist = doc_settings.persist,
    button_text = doc_settings.button_text
  }
  
  -- Override with cell-level attributes
  if has_attr(el, "toggle") then
    settings.toggle = get_bool_attr(el, "toggle", settings.toggle)
  end
  
  if has_attr(el, "output-hidden") then
    settings.hidden = get_bool_attr(el, "output-hidden", settings.hidden)
  end
  
  if has_attr(el, "output-sync") then
    settings.sync = get_bool_attr(el, "output-sync", settings.sync)
  end
  
  if has_attr(el, "persist") then
    settings.persist = get_bool_attr(el, "persist", settings.persist)
  end
  
  if has_attr(el, "button-text") then
    settings.button_text = get_string_attr(el, "button-text", settings.button_text)
  end
  
  return settings
end

-- ============================================================================
-- Class Application
-- ============================================================================

--- Add the base toggleable class to an element.
-- @param el table The Pandoc Div element
-- @return nil
local function add_toggleable_class(el)
  el.classes:insert("toggleable-cell")
end

--- Add sync mode class to an element.
-- @param el table The Pandoc Div element
-- @param is_sync boolean Whether sync mode is enabled
-- @return nil
local function add_sync_class(el, is_sync)
  if is_sync then
    el.classes:insert("output-sync-on")
  else
    el.classes:insert("output-sync-off")
  end
end

--- Add initially hidden class to an element.
-- @param el table The Pandoc Div element
-- @param is_hidden boolean Whether output should be hidden
-- @return nil
local function add_hidden_class(el, is_hidden)
  if is_hidden then
    el.classes:insert("initially-hidden")
  end
end

--- Add persistence class to an element.
-- @param el table The Pandoc Div element
-- @param is_persist boolean Whether persistence is enabled
-- @return nil
local function add_persist_class(el, is_persist)
  if is_persist then
    el.classes:insert("persist-enabled")
  end
end

--- Add button text data attribute to an element.
-- @param el table The Pandoc Div element
-- @param button_text string The button text
-- @return nil
local function add_button_text_attr(el, button_text)
  if button_text ~= DEFAULT_SETTINGS.button_text then
    el.attributes["data-button-text"] = button_text
  end
end

--- Apply all toggle classes and attributes to an element.
-- @param el table The Pandoc Div element
-- @param settings table The cell settings
-- @return nil
local function apply_toggle_classes(el, settings)
  add_toggleable_class(el)
  add_sync_class(el, settings.sync)
  add_hidden_class(el, settings.hidden)
  add_persist_class(el, settings.persist)
  add_button_text_attr(el, settings.button_text)
end

-- ============================================================================
-- HTML Dependency Injection
-- ============================================================================

--- Build the list of HTML classes for document-level features.
-- @return table Array of class names
local function build_html_classes()
  local classes = {}
  
  if doc_settings.global_toggle then
    table.insert(classes, "toggle-global-enabled")
  end
  
  if doc_settings.persist then
    table.insert(classes, "toggle-persist-enabled")
  end
  
  return classes
end

--- Build the JavaScript for setting data attributes.
-- @return string JavaScript code or empty string
local function build_attr_script()
  if doc_settings.button_text ~= DEFAULT_SETTINGS.button_text then
    return string.format(
      "document.documentElement.setAttribute('data-toggle-global-text', %q);",
      doc_settings.button_text
    )
  end
  return ""
end

--- Generate the header injection script.
-- Creates a script tag that adds classes to the HTML element.
-- @param classes table Array of class names
-- @param attr_script string Additional attribute script
-- @return string The complete script tag
local function generate_injection_script(classes, attr_script)
  local class_str = table.concat(classes, " ")
  
  return string.format([[
<script>
(function() {
  var classes = %q.split(' ').filter(Boolean);
  classes.forEach(function(c) { document.documentElement.classList.add(c); });
  %s
})();
</script>
]], class_str, attr_script)
end

--- Add toggle HTML dependency to the document.
-- Registers CSS and JS files with Quarto.
-- @return nil
local function add_html_dependency()
  quarto.doc.add_html_dependency({
    name = "toggle",
    version = "0.1.0",
    scripts = {"toggle.js"},
    stylesheets = {"toggle.css"}
  })
end

--- Inject document-level configuration into the HTML.
-- Adds classes and attributes to the HTML element.
-- @return nil
local function inject_document_config()
  local classes = build_html_classes()
  local attr_script = build_attr_script()
  
  if #classes > 0 or attr_script ~= "" then
    local script = generate_injection_script(classes, attr_script)
    quarto.doc.include_text("in-header", script)
  end
end

-- ============================================================================
-- Format Checking
-- ============================================================================

--- Check if the current output format is HTML.
-- @return boolean True if output format is HTML
local function is_html_format()
  return quarto.doc.is_format("html")
end

--- Check if an element is a code cell.
-- @param el table The Pandoc Div element
-- @return boolean True if element has "cell" class
local function is_cell(el)
  return el.classes:includes("cell")
end

-- ============================================================================
-- Toggle ID Marking
-- ============================================================================

--- Check if a CodeBlock element is a code block within a cell.
-- @param el table The Pandoc element
-- @return boolean True if element is a cell-code section
local function is_code_block(el)
  if el.t ~= "CodeBlock" then
    return false
  end
  return el.classes:includes("cell-code")
end

--- Check if a Div element is an output section.
-- @param el table The Pandoc element
-- @return boolean True if element is a cell-output section
local function is_output_block(el)
  if el.t ~= "Div" then
    return false
  end
  for _, class in ipairs(el.classes) do
    if class:match("^cell%-output") then
      return true
    end
  end
  return false
end

--- Generate a toggle ID for a code-output pair.
-- @param cell_idx number The cell index
-- @param pair_idx number The pair index within the cell
-- @return string The generated toggle ID
local function generate_toggle_id(cell_idx, pair_idx)
  return string.format("toggle-%d-%d", cell_idx, pair_idx)
end

--- Find and mark code-output pairs within a cell.
-- Traverses cell content recursively to find code blocks and their associated outputs,
-- assigning matching toggle IDs to each pair.
-- @param el table The cell Div element
-- @param cell_idx number The cell index
-- @return table The modified cell with marked elements
local function mark_toggle_pairs(el, cell_idx)
  local pair_idx = 0
  local current_code_id = nil

  -- Walk through all elements in the cell
  el = el:walk({
    -- Handle CodeBlock elements (code sections)
    CodeBlock = function(item)
      if is_code_block(item) then
        -- Found a code block, generate new ID
        pair_idx = pair_idx + 1
        current_code_id = generate_toggle_id(cell_idx, pair_idx)
        item.attributes["data-toggle-id"] = current_code_id
        return item
      end
      return nil
    end,

    -- Handle Div elements (output sections)
    Div = function(item)
      if is_output_block(item) then
        -- Found an output block
        if current_code_id then
          -- Mark with the current code block's ID
          item.attributes["data-toggle-output"] = current_code_id
        end
        return item
      end
      return nil
    end
  })

  return el
end

-- ============================================================================
-- Filter Functions
-- ============================================================================

--- Process document metadata to extract toggle settings.
-- First filter pass: reads YAML configuration.
-- @param meta table The document metadata from Pandoc
-- @return table The metadata table (unchanged)
local function process_meta(meta)
  doc_settings = extract_settings(meta)
  
  if doc_settings.toggle_enabled or doc_settings.global_toggle then
    needs_toggle = true
  end
  
  return meta
end

--- Process a single cell Div element.
-- Determines if toggle should be applied and adds necessary classes.
-- @param el table The Pandoc Div element
-- @return table The modified Div element
local function process_cell(el)
  local settings = resolve_cell_settings(el)

  -- Track if cell-level toggle was enabled
  if has_attr(el, "toggle") and settings.toggle then
    needs_toggle = true
  end

  -- Skip if toggle not enabled for this cell
  if not settings.toggle then
    return el
  end

  -- Apply toggle classes to the cell
  apply_toggle_classes(el, settings)

  -- Mark code-output pairs with matching IDs
  el = mark_toggle_pairs(el, cell_index)

  -- Increment cell counter for next cell
  cell_index = cell_index + 1

  return el
end

--- Process Div elements to add toggle functionality.
-- Second filter pass: processes each cell div.
-- @param el table The Pandoc Div element
-- @return table The modified Div element
local function process_div(el)
  if not is_html_format() then
    return el
  end
  
  if not is_cell(el) then
    return el
  end
  
  return process_cell(el)
end

--- Finalize and add HTML dependencies.
-- Third filter pass: adds CSS/JS dependencies and injects config.
-- @param meta table The document metadata from Pandoc
-- @return table The metadata table (unchanged)
local function finalize_meta(meta)
  if not needs_toggle then
    return meta
  end
  
  add_html_dependency()
  inject_document_config()
  
  return meta
end

-- ============================================================================
-- Filter Specification
-- ============================================================================

--- Filter specification for Quarto/Pandoc.
-- Defines processing order: Meta -> Div -> Meta (finalize).
-- @return table The filter specification
return {
  { Meta = process_meta },
  { Div = process_div },
  { Meta = finalize_meta }
}

File: _extensions/toggle/toggle.js
/**
 * Toggle Extension for Quarto
 * Provides toggle buttons to show/hide code output sections
 *
 * @author James J Balamuta
 * @version 0.2.0
 */

(function() {
  'use strict';

  // ===========================================================================
  // Configuration
  // ===========================================================================

  /**
   * Central configuration object for class names, selectors, and defaults
   */
  var CONFIG = {
    classes: {
      toggleBtn: 'code-toggle-btn',
      outputHidden: 'output-hidden',
      initiallyHidden: 'initially-hidden',
      syncOn: 'output-sync-on',
      hidden: 'hidden',
      globalBtn: 'toggle-global-btn',
      allHidden: 'all-hidden',
      persistEnabled: 'persist-enabled'
    },
    selectors: {
      toggleableCell: '.toggleable-cell',
      toggleId: '[data-toggle-id]',
      toggleOutput: '[data-toggle-output]',
      globalEnabled: 'toggle-global-enabled',
      persistEnabled: 'toggle-persist-enabled'
    },
    storageKeyPrefix: 'quarto-toggle-',
    defaultButtonText: 'Output',
    tooltips: {
      show: 'Show output',
      hide: 'Hide output',
      showAll: 'Show all outputs',
      hideAll: 'Hide all outputs'
    }
  };

  // ===========================================================================
  // Storage Module
  // ===========================================================================

  /**
   * Handles localStorage operations for persisting toggle state
   */
  var Storage = {
    _available: null,

    /**
     * Check if localStorage is available (cached after first check)
     * @returns {boolean} True if localStorage is available
     */
    isAvailable: function() {
      if (this._available !== null) return this._available;
      try {
        var test = '__storage_test__';
        localStorage.setItem(test, test);
        localStorage.removeItem(test);
        this._available = true;
      } catch (e) {
        this._available = false;
      }
      return this._available;
    },

    /**
     * Generate a storage key for a specific item based on page path
     * @param {string} suffix - The suffix to append to the key
     * @returns {string} The full storage key
     */
    createKey: function(suffix) {
      var pageId = window.location.pathname.replace(/\//g, '-');
      return CONFIG.storageKeyPrefix + pageId + '-' + suffix;
    },

    /**
     * Save a value to localStorage
     * @param {string} key - The storage key
     * @param {*} value - The value to store (will be JSON stringified)
     */
    save: function(key, value) {
      if (!this.isAvailable()) return;
      try {
        localStorage.setItem(key, JSON.stringify(value));
      } catch (e) {
        // Silently fail (quota exceeded, etc.)
      }
    },

    /**
     * Load a value from localStorage
     * @param {string} key - The storage key
     * @returns {*} The parsed value, or null if not found
     */
    load: function(key) {
      if (!this.isAvailable()) return null;
      try {
        var stored = localStorage.getItem(key);
        return stored !== null ? JSON.parse(stored) : null;
      } catch (e) {
        return null;
      }
    }
  };

  // ===========================================================================
  // Accessibility Module
  // ===========================================================================

  /**
   * Handles accessibility features including screen reader announcements
   */
  var A11y = {
    liveRegion: null,

    /**
     * Initialize the ARIA live region for screen reader announcements
     */
    init: function() {
      if (this.liveRegion) return;
      this.liveRegion = document.createElement('div');
      this.liveRegion.setAttribute('role', 'status');
      this.liveRegion.setAttribute('aria-live', 'polite');
      this.liveRegion.setAttribute('aria-atomic', 'true');
      this.liveRegion.className = 'sr-only';
      // Visually hidden but accessible to screen readers
      this.liveRegion.style.cssText =
        'position:absolute;width:1px;height:1px;padding:0;margin:-1px;' +
        'overflow:hidden;clip:rect(0,0,0,0);white-space:nowrap;border:0;';
      document.body.appendChild(this.liveRegion);
    },

    /**
     * Announce a message to screen readers
     * @param {string} message - The message to announce
     */
    announce: function(message) {
      if (!this.liveRegion) this.init();
      // Clear and re-set to ensure announcement is triggered
      this.liveRegion.textContent = '';
      requestAnimationFrame(function() {
        A11y.liveRegion.textContent = message;
      });
    }
  };

  // ===========================================================================
  // Button Module
  // ===========================================================================

  /**
   * Handles toggle button creation and state management
   */
  var Button = {
    /**
     * Update button state (classes, ARIA attributes, tooltip)
     * @param {HTMLButtonElement} btn - The button element
     * @param {boolean} isHidden - Whether the associated output is hidden
     */
    updateState: function(btn, isHidden) {
      btn.classList.toggle(CONFIG.classes.outputHidden, isHidden);
      btn.setAttribute('aria-expanded', String(!isHidden));
      btn.setAttribute('aria-label', isHidden ? 'Show output' : 'Hide output');
      btn.setAttribute('data-tooltip', isHidden ? CONFIG.tooltips.show : CONFIG.tooltips.hide);
    },

    /**
     * Create a new toggle button
     * @param {string} pairId - The ID linking this button to its output(s)
     * @param {string} buttonText - The button label text
     * @param {boolean} isHidden - Initial hidden state
     * @returns {HTMLButtonElement} The created button
     */
    create: function(pairId, buttonText, isHidden) {
      var btn = document.createElement('button');
      btn.className = CONFIG.classes.toggleBtn;
      btn.type = 'button';
      btn.tabIndex = 0;
      btn.dataset.pair = pairId;
      btn.textContent = buttonText || CONFIG.defaultButtonText;
      this.updateState(btn, isHidden);
      this.addKeyboardSupport(btn);
      return btn;
    },

    /**
     * Add keyboard support (Enter/Space to activate)
     * @param {HTMLButtonElement} btn - The button element
     */
    addKeyboardSupport: function(btn) {
      btn.addEventListener('keydown', function(e) {
        if (e.key === 'Enter' || e.key === ' ') {
          e.preventDefault();
          btn.click();
        }
      });
    }
  };

  // ===========================================================================
  // Toggle Module
  // ===========================================================================

  /**
   * Core toggle logic for showing/hiding outputs
   */
  var Toggle = {
    /**
     * Set hidden state on multiple output elements
     * @param {HTMLElement[]} outputs - Array of output elements
     * @param {boolean} isHidden - Whether to hide the outputs
     */
    setOutputsHidden: function(outputs, isHidden) {
      for (var i = 0; i < outputs.length; i++) {
        outputs[i].classList.toggle(CONFIG.classes.hidden, isHidden);
      }
    },

    /**
     * Update state on multiple buttons
     * @param {HTMLButtonElement[]} buttons - Array of button elements
     * @param {boolean} isHidden - Whether outputs are hidden
     */
    updateButtons: function(buttons, isHidden) {
      for (var i = 0; i < buttons.length; i++) {
        Button.updateState(buttons[i], isHidden);
      }
    },

    /**
     * Toggle a single output (or group of outputs for one code block)
     * @param {HTMLElement[]} outputs - The output elements to toggle
     * @param {HTMLButtonElement} btn - The button that triggered the toggle
     * @param {string|null} storageKey - Optional key for persisting state
     */
    single: function(outputs, btn, storageKey) {
      var isCurrentlyHidden = outputs[0].classList.contains(CONFIG.classes.hidden);
      var newState = !isCurrentlyHidden;

      this.setOutputsHidden(outputs, newState);
      Button.updateState(btn, newState);
      A11y.announce(newState ? 'Output hidden' : 'Output shown');

      if (storageKey) {
        Storage.save(storageKey, newState);
      }
    },

    /**
     * Toggle all outputs in a synced group or globally
     * @param {HTMLElement[]} outputs - All output elements to toggle
     * @param {HTMLButtonElement[]} buttons - All buttons to update
     * @param {string|null} storageKey - Optional key for persisting state
     * @returns {boolean} The new hidden state
     */
    all: function(outputs, buttons, storageKey) {
      // Check if all are currently hidden
      var allHidden = true;
      for (var i = 0; i < outputs.length; i++) {
        if (!outputs[i].classList.contains(CONFIG.classes.hidden)) {
          allHidden = false;
          break;
        }
      }

      var newState = !allHidden;

      this.setOutputsHidden(outputs, newState);
      this.updateButtons(buttons, newState);
      A11y.announce(newState ? 'All outputs hidden' : 'All outputs shown');

      if (storageKey) {
        Storage.save(storageKey, newState);
      }

      return newState;
    }
  };

  // ===========================================================================
  // Cell Processor
  // ===========================================================================

  /**
   * Processes toggleable cells and sets up their toggle functionality
   */
  var CellProcessor = {
    /**
     * Process a single toggleable cell
     * @param {HTMLElement} cell - The cell container element
     * @param {number} cellIndex - The index of this cell (for storage keys)
     */
    process: function(cell, cellIndex) {
      var shouldHideInitially = cell.classList.contains(CONFIG.classes.initiallyHidden);
      var outputSyncOn = cell.classList.contains(CONFIG.classes.syncOn);
      var persistEnabled = cell.classList.contains(CONFIG.classes.persistEnabled);
      var customButtonText = cell.dataset.buttonText || CONFIG.defaultButtonText;

      var cellStorageKey = persistEnabled ? Storage.createKey('cell-' + cellIndex) : null;
      var persistedState = cellStorageKey ? Storage.load(cellStorageKey) : null;

      var codeBlocks = cell.querySelectorAll(CONFIG.selectors.toggleId);
      if (codeBlocks.length === 0) return;

      var allOutputs = [];
      var allButtons = [];
      var self = this;

      codeBlocks.forEach(function(codeBlock) {
        var toggleId = codeBlock.getAttribute('data-toggle-id');
        var outputs = cell.querySelectorAll('[data-toggle-output="' + toggleId + '"]');

        if (outputs.length === 0) return;

        var pairStorageKey = persistEnabled ? Storage.createKey('pair-' + toggleId) : null;
        var initiallyHidden = self.determineInitialState(
          shouldHideInitially,
          persistEnabled,
          outputSyncOn,
          persistedState,
          pairStorageKey
        );

        var toggleBtn = Button.create(toggleId, customButtonText, initiallyHidden);
        codeBlock.appendChild(toggleBtn);

        Toggle.setOutputsHidden(outputs, initiallyHidden);

        if (outputSyncOn) {
          self.collectForSync(outputs, toggleBtn, allOutputs, allButtons);
        } else {
          self.setupIndividualHandler(outputs, toggleBtn, pairStorageKey);
        }
      });

      // Set up sync mode handlers if enabled
      if (outputSyncOn && allButtons.length > 0) {
        this.setupSyncHandlers(allOutputs, allButtons, cellStorageKey);
      }
    },

    /**
     * Determine the initial hidden state based on config and persisted state
     * @param {boolean} shouldHideInitially - Document/cell default
     * @param {boolean} persistEnabled - Whether persistence is enabled
     * @param {boolean} outputSyncOn - Whether sync mode is enabled
     * @param {boolean|null} persistedState - Cell-level persisted state
     * @param {string|null} pairStorageKey - Key for pair-level persistence
     * @returns {boolean} The initial hidden state
     */
    determineInitialState: function(shouldHideInitially, persistEnabled, outputSyncOn, persistedState, pairStorageKey) {
      var initiallyHidden = shouldHideInitially;

      if (persistEnabled) {
        // Use cell-level key for sync mode, pair-level for individual
        var savedState = outputSyncOn ? persistedState : Storage.load(pairStorageKey);
        if (savedState !== null) {
          initiallyHidden = savedState;
        }
      }

      return initiallyHidden;
    },

    /**
     * Collect outputs and button for sync mode
     * @param {NodeList} outputs - Output elements for this code block
     * @param {HTMLButtonElement} toggleBtn - The toggle button
     * @param {HTMLElement[]} allOutputs - Accumulator for all outputs
     * @param {HTMLButtonElement[]} allButtons - Accumulator for all buttons
     */
    collectForSync: function(outputs, toggleBtn, allOutputs, allButtons) {
      for (var i = 0; i < outputs.length; i++) {
        allOutputs.push(outputs[i]);
      }
      allButtons.push(toggleBtn);
    },

    /**
     * Set up click handler for individual (non-synced) mode
     * @param {NodeList} outputs - Output elements for this code block
     * @param {HTMLButtonElement} toggleBtn - The toggle button
     * @param {string|null} storageKey - Optional persistence key
     */
    setupIndividualHandler: function(outputs, toggleBtn, storageKey) {
      var outputsArray = Array.from(outputs);
      toggleBtn.addEventListener('click', function(e) {
        e.stopPropagation();
        Toggle.single(outputsArray, toggleBtn, storageKey);
      });
    },

    /**
     * Set up click handlers for sync mode (all buttons control all outputs)
     * @param {HTMLElement[]} allOutputs - All output elements in the cell
     * @param {HTMLButtonElement[]} allButtons - All toggle buttons in the cell
     * @param {string|null} storageKey - Optional persistence key
     */
    setupSyncHandlers: function(allOutputs, allButtons, storageKey) {
      for (var i = 0; i < allButtons.length; i++) {
        (function(outputs, buttons, key) {
          buttons[i].addEventListener('click', function(e) {
            e.stopPropagation();
            Toggle.all(outputs, buttons, key);
          });
        })(allOutputs, allButtons, storageKey);
      }
    }
  };

  // ===========================================================================
  // Global Toggle
  // ===========================================================================

  /**
   * Handles the global toggle button that controls all outputs on the page
   */
  var GlobalToggle = {
    /**
     * Initialize the global toggle button if enabled
     */
    init: function() {
      var docElement = document.documentElement;
      if (!docElement.classList.contains(CONFIG.selectors.globalEnabled)) return;
      if (document.querySelector('.' + CONFIG.classes.globalBtn)) return;

      var allOutputs = Array.from(document.querySelectorAll(CONFIG.selectors.toggleOutput));
      var allToggleButtons = Array.from(document.querySelectorAll('.' + CONFIG.classes.toggleBtn));

      if (allOutputs.length === 0) return;

      var persistEnabled = docElement.classList.contains(CONFIG.selectors.persistEnabled);
      var globalStorageKey = persistEnabled ? Storage.createKey('global') : null;
      var buttonText = docElement.dataset.toggleGlobalText || 'Toggle All Outputs';

      var globalBtn = this.createButton(buttonText);

      this.restoreState(globalStorageKey, allOutputs, allToggleButtons, globalBtn);
      this.setupHandler(globalBtn, allOutputs, allToggleButtons, globalStorageKey);

      document.body.appendChild(globalBtn);
    },

    /**
     * Create the global toggle button element
     * @param {string} text - The button label text
     * @returns {HTMLButtonElement} The created button
     */
    createButton: function(text) {
      var btn = document.createElement('button');
      btn.className = CONFIG.classes.globalBtn;
      btn.type = 'button';
      btn.tabIndex = 0;
      btn.textContent = text;
      this.updateButtonState(btn, false);
      Button.addKeyboardSupport(btn);
      return btn;
    },

    /**
     * Update the global button state
     * @param {HTMLButtonElement} btn - The global button
     * @param {boolean} allHidden - Whether all outputs are hidden
     */
    updateButtonState: function(btn, allHidden) {
      btn.classList.toggle(CONFIG.classes.allHidden, allHidden);
      var label = allHidden ? CONFIG.tooltips.showAll : CONFIG.tooltips.hideAll;
      btn.setAttribute('aria-label', label);
      btn.title = label;
    },

    /**
     * Restore persisted global state on page load
     * @param {string|null} storageKey - The persistence key
     * @param {HTMLElement[]} outputs - All output elements
     * @param {HTMLButtonElement[]} buttons - All toggle buttons
     * @param {HTMLButtonElement} globalBtn - The global button
     */
    restoreState: function(storageKey, outputs, buttons, globalBtn) {
      if (!storageKey) return;

      var savedState = Storage.load(storageKey);
      if (savedState === null) return;

      Toggle.setOutputsHidden(outputs, savedState);
      Toggle.updateButtons(buttons, savedState);
      this.updateButtonState(globalBtn, savedState);
    },

    /**
     * Set up click handler for the global button
     * @param {HTMLButtonElement} globalBtn - The global button
     * @param {HTMLElement[]} allOutputs - All output elements
     * @param {HTMLButtonElement[]} allToggleButtons - All toggle buttons
     * @param {string|null} storageKey - Optional persistence key
     */
    setupHandler: function(globalBtn, allOutputs, allToggleButtons, storageKey) {
      var self = this;

      globalBtn.addEventListener('click', function() {
        var newState = Toggle.all(allOutputs, allToggleButtons, storageKey);
        self.updateButtonState(globalBtn, newState);
      });
    }
  };

  // ===========================================================================
  // Initialization
  // ===========================================================================

  /**
   * Initialize the toggle extension
   */
  function init() {
    A11y.init();

    // Process all toggleable cells
    var toggleableCells = document.querySelectorAll(CONFIG.selectors.toggleableCell);
    toggleableCells.forEach(function(cell, index) {
      CellProcessor.process(cell, index);
    });

    // Initialize global toggle if enabled
    GlobalToggle.init();
  }

  // Run when DOM is ready
  if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', init);
  } else {
    init();
  }
})();
